lib/Dialect/AIEX/IR/AIEXDialect.cpp

```diff
@@ -157,41 +157,70 @@ LogicalResult AIEX::NpuDmaMemcpyNdOp::verify() {
-    return emitOpError("Size 3 exceeds the [1:64] range.");
-  if (strides[1] && sizes[1] > 0x3FF)
-    return emitOpError("Size 1 exceeds the [0:1023] range.");
-  if (strides[0] && sizes[0] > 0x3FF)
-    return emitOpError("Size 0 exceeds the [0:1023] range.");
-  if (strides[2] > 0x100000)
-    return emitOpError("Stride 3 exceeds the [1:1M] range.");
-  if (strides[1] > 0x100000)
-    return emitOpError("Stride 2 exceeds the [1:1M] range.");
-  if (strides[0] > 0x100000)
-    return emitOpError("Stride 1 exceeds the [1:1M] range.");
+  // The experimental HSA target uses this op on AIE1, skip all the AIE2
+  // specific checks
+  if (targetModel.getTargetArch() == AIE::AIEArch::AIE1)
+    return success();
+
+  uint32_t wrap_bits = 0;
+  uint32_t step_bits = 0;
+  uint32_t iter_bits = 6;
+  if (targetModel.isShimNOCTile(getX(), getY())) {
+    step_bits = 20; // XAIEMLGBL_NOC_MODULE_DMA_BD0_3_D0_STEPSIZE_WIDTH
+    wrap_bits = 10; // XAIEMLGBL_NOC_MODULE_DMA_BD0_3_D0_WRAP_WIDTH
+  } else if (targetModel.isMemTile(getX(), getY())) {
+    step_bits = 17; // XAIEMLGBL_MEM_TILE_MODULE_DMA_BD0_2_D0_STEPSIZE_WIDTH
+    wrap_bits = 10; // XAIEMLGBL_MEM_TILE_MODULE_DMA_BD0_2_D0_WRAP_WIDTH
+  } else if (targetModel.isCoreTile(getX(), getY())) {
+    step_bits = 13; // XAIEMLGBL_MEMORY_MODULE_DMA_BD0_2_D0_STEPSIZE_WIDTH
+    wrap_bits = 8;  // XAIEMLGBL_MEMORY_MODULE_DMA_BD0_3_D0_WRAP_WIDTH
+  } else {
+    return emitOpError("Unsupported tile type at (" + std::to_string(getX()) +
+                       ", " + std::to_string(getY()) +
+                       ") Must be ShimNOC, Mem or Core.");
+  }
+
+  if (sizes[3] > (1 << iter_bits))
+    return emitOpError(
+        "Size 3 exceeds the [1:" + std::to_string(1 << iter_bits) + "] range.");
+  if (strides[2] && sizes[1] > (1 << wrap_bits) - 1)
+    return emitOpError("Size 1 exceeds the [0:" +
+                       std::to_string((1 << wrap_bits) - 1) + "] range.");
+  if (strides[1] && sizes[0] > (1 << wrap_bits) - 1)
+    return emitOpError("Size 0 exceeds the [0:" +
+                       std::to_string((1 << wrap_bits) - 1) + "] range.");
+  if (strides[3] > (1 << step_bits))
+    return emitOpError("Stride 3 exceeds the [1:" +
+                       std::to_string(1 << step_bits) + "] range.");
+  if (strides[2] > (1 << step_bits))
+    return emitOpError("Stride 2 exceeds the [1:" +
+                       std::to_string(1 << step_bits) + "] range.");
+  if (strides[1] > (1 << step_bits))
+    return emitOpError("Stride 1 exceeds the [1:" +
+                       std::to_string(1 << step_bits) + "] range.");
 
   if (offset % 4 != 0) {
     return emitOpError("Offset must be 4-byte-aligned.");
   }
 
-  bool error = false;
-  std::stringstream msg;
-  for (int i = 0; i < 3; i++) {
+  for (int i = 0; i < 4; i++) {
+    // strides[0] == 1 is ok iff the tranfer size is a multiple of
+    // addressGranularity, which is checked below
+    if (i == 0 && raw_strides[i] == 1)
+      continue;
```

Comment:
Looks good. Technically something like
```
sizes=[addressGranularity/2, addressGranularity/2]
strides=[addressGranularity/2, 1]
```
(and similar examples for higher dimensions, with addressGranularity in bytes)
is equivalent to
```
sizes=[addressGranularity]
strides=[1]
```
but would currently error.

I think it's a non-issue as this is clearly a contrived example with no real uses. (Except maybe automatically generated code?)

So this looks good to me with this check.